java-conductr-bundle-lib/src/main/java/com/typesafe/conductr/bundlelib/java/LocationCache.java

```diff
@@ -24,23 +24,34 @@
 
     @Override
     public CompletionStage<Optional<URI>> getOrElseUpdate(String serviceName, Supplier<CompletionStage<Optional<Tuple<URI, Optional<Duration>>>>> op) {
-        return cache.computeIfAbsent(serviceName, sn ->
-            op
-                .get()
-                .whenCompleteAsync((locationAndMaxAge, e) -> {
-                    try {
-                        Duration maxAge = locationAndMaxAge.get()._2.get();
-                        reaperTimer.schedule(new TimerTask() {
-                            @Override
-                            public void run() {
+        return cache
+                .computeIfAbsent(serviceName, sn ->
+                    op.get()
+                        .whenCompleteAsync((locationAndMaxAge, e) -> {
+                            try {
+                                Duration maxAge = locationAndMaxAge.get()._2.get();
+                                reaperTimer.schedule(new TimerTask() {
+                                    @Override
+                                    public void run() {
+                                        cache.remove(sn);
+                                    }
+                                }, maxAge.toMillis());
+                            } catch (NoSuchElementException nse) {
                                 cache.remove(sn);
```

Comment:
That remove could still potentially complete before the `computeIfAbsent` completes.